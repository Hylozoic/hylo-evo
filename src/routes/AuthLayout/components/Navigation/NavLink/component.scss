.item {
  // FIXME this makes padding of 12px, actual spacing is 12.5px
  padding: 0 $space-3x;
  // height and line-height maybe should use spacing units too?
  height: 40px;
  line-height: 40px;
  border-radius: 5px;

}

.link {
  display: flex;
  align-items: center;
  color: $color-rhino-60;
  text-decoration: none;
  &:hover {
    text-decoration: none;
    color: $color-rhino;
  }

  transition: 1.5s;

  &.collapsed {
    color: transparent;
  }
}



.active {
  background-color: $color-iceberg;
  .link {
    text-decoration: none;
    color: $color-rhino;
    &.collapsed {
      color: transparent;
    }
  }
}

.icon {
  margin-right: 24px;
}

.label {
  min-width: 102px;
}
