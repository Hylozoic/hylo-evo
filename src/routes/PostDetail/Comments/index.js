import Comments from './Comments'
import connector from './Comments.connector'
export default connector(Comments)
