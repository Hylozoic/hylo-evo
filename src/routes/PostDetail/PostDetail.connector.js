import { connect } from 'react-redux'
import { push } from 'react-router-redux'
import { postUrl } from 'util/index'
import fetchPost from 'store/actions/fetchPost'
import getParam from 'store/selectors/getParam'
import getPost from 'store/selectors/getPost'

export function mapStateToProps (state, props) {
  const slug = getParam('slug', state, props)
  return {
    post: getPost(state, props),
    id: getParam('postId', state, props),
    slug,
    showCommunity: !slug
  }
}

export const mapDispatchToProps = (dispatch, props) => {
  const { location } = props

  const removePostDetailFromPath = pathname =>
    pathname.replace(/\/p\/(.+)/, '')

  const closeLocation = {
    ...props.location,
    pathname: removePostDetailFromPath(location.pathname)
  }

  const postId = getParam('postId', {}, props)

  return {
    fetchPost: () => dispatch(fetchPost(getParam('postId', {}, props))),
    onClose: () => dispatch(push(closeLocation)),
    editPost: () => dispatch(push(postUrl(postId) + '/edit'))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
