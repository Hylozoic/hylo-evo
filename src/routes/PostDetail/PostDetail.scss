.post {
  padding-top: $space-6x;
  background-color: white;
}

.header {
  margin-bottom: 10px;
}

.header-sticky {
  position: fixed;
  top: $topnav-height;
  background-image: linear-gradient(0deg, rgba(255,255,255,0.00) 0%, $color-white 25%);
  padding: $space-6x 0;
  z-index: 1;
}

.activity-sticky {
  z-index: 1;
  position: fixed;
  top: $topnav-height;
  background: $color-white;
}

.image {
  background-position: center;
  background-size: cover;
  @include fluid-height(62%);
  width: 100%;
  margin-bottom: 10px;
}

.body {
  padding-left: 20px;
  padding-right: 20px;
}

.activity-header {
  composes: tag from 'css/typography.scss';
  padding-bottom: 8px;
  padding-left: 27px;
}

.join-project-button-container {
  width: 100%;
  text-align: center;
  margin-bottom: 12px;
}

.join-project-button {
  margin: 0px;
}

// ProjectContributions

.project-contributions {
  width: 100%;
  border-top: 1px solid $color-cape-cod-10;  
  padding: 10px 16px;
}

.project-contributions-total {
  composes: caption-lt-lg from 'css/typography.scss';
  margin-top: 10px;
}

.amount-row {
  display: flex;
  align-items: center;
  margin-bottom: 5px;
}

.amount-label {
  composes: caption-lt-lg from 'css/typography.scss';
  margin-right: 10px;
}

.success-notification {
  color: white;
  background-color: $color-mountain-meadow;
  padding: 5px;
  border-radius: 4px;
  margin-bottom: 10px;
}

.error-notification {
  color: white;
  background-color: $color-amaranth;
  padding: 5px;
  border-radius: 4px;
  margin-bottom: 10px;
}

.cancel-button {
  margin-left: 10px;
}

// PostTag

.tags {
  padding: 0px 25px;
  margin-bottom: 10px;
}

.tag {
  // TODO: Extract typography
  @include font-regular;
  color: $color-caribbean-green;
  font-size: 12px;
  &:hover {
    text-decoration: none;
    color: $color-gossamer;
  }
  margin-right: 7px;
}
