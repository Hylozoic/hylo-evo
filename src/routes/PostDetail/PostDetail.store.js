import { FETCH_POST } from 'store/constants'

export function fetchPost (id, opts = {}) {
  return {
    type: FETCH_POST,
    graphql: {
      query: `query ($id: ID) {
        post(id: $id) {
          id
          title
          details
          type
          creator {
            id
            name
            avatarUrl
          }
          createdAt
          updatedAt
          commenters(first: 2) {
            id
            name
            avatarUrl
          }
          commentersTotal
          comments(first: 10, order: "desc") {
            items {
              id
              text
              creator {
                id
                name
                avatarUrl
              }
              createdAt
            }
            total
            hasMore
          }
          linkPreview {
            title
            url
            imageUrl
          }
          votesTotal
          communities {
            id
            name
            slug
          }
        }
      }`,
      variables: {
        id
      }
    },
    meta: {
      extractModel: 'Post'
    }
  }
}
