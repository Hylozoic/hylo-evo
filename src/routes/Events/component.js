import React from 'react'
import { Link } from 'react-router-dom'
import './component.scss'

const { string, bool } = React.PropTypes

export default function Events () {
  return <div styleName='event-list'>
    <div><Link to='/events/1'>Event 1</Link></div>
    <div><Link to='/events/2'>Event 2</Link></div>
    <div><Link to='/events/3'>Event 3</Link></div>
    <div><Link to='/events/4'>Event 4</Link></div>
  </div>
}
Events.propTypes = {
  label: string,
  color: string,
  active: bool,
  className: string
}
