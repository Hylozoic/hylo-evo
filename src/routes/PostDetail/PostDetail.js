import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { throttle, isEmpty } from 'lodash/fp'
import './PostDetail.scss'
import { PostImage, PostBody, PostFooter, PostHeader } from 'components/PostCard'
import ScrollListener from 'components/ScrollListener'
import Comments from './Comments'
import { tagUrl } from 'util/index'
import { DETAIL_COLUMN_ID, position } from 'util/scrolling'
import SocketSubscriber from 'components/SocketSubscriber'
import Loading from 'components/Loading'
import NotFound from 'components/NotFound'

const { bool, func, object, string } = PropTypes

// the height of the header plus the padding-top
const STICKY_HEADER_SCROLL_OFFSET = 78

export default class PostDetail extends Component {
  static propTypes = {
    post: object,
    id: string,
    currentUser: object,
    slug: string,
    fetchPost: func,
    showCommunity: bool
  }

  constructor (props) {
    super(props)
    this.state = {
      atHeader: false,
      headerWidth: 0,
      headerScrollOffset: 0,
      atActivity: false,
      activityWidth: 0,
      activityScrollOffset: 0
    }
  }

  setHeaderStateFromDOM = () => {
    const container = document.getElementById(DETAIL_COLUMN_ID)
    if (!container) return
    this.setState({
      headerWidth: container.offsetWidth
    })
  }

  setActivityStateFromDOM = activity => {
    const element = ReactDOM.findDOMNode(activity)
    const container = document.getElementById(DETAIL_COLUMN_ID)
    if (!element || !container) return
    const offset = position(element, container).y - STICKY_HEADER_SCROLL_OFFSET
    this.setState({
      activityWidth: element.offsetWidth,
      activityScrollOffset: offset
    })
  }

  componentDidMount () {
    this.onPostIdChange()
  }

  componentDidUpdate (prevProps) {
    if (this.props.id && this.props.id !== prevProps.id) {
      this.onPostIdChange()
    }
  }

  onPostIdChange = () => {
    this.props.fetchPost()
  }

  handleScroll = throttle(100, event => {
    const { scrollTop } = event.target
    const {
      atHeader,
      atActivity,
      headerScrollOffset,
      activityScrollOffset
    } = this.state
    if (!atActivity && scrollTop >= activityScrollOffset) {
      this.setState({atActivity: true})
    } else if (atActivity && scrollTop < activityScrollOffset) {
      this.setState({atActivity: false})
    }

    if (!atHeader && scrollTop > headerScrollOffset) {
      this.setState({atHeader: true})
    } else if (atHeader && scrollTop <= headerScrollOffset) {
      this.setState({atHeader: false})
    }
  })

  render () {
    const { post, slug, pending } = this.props
    const { atHeader, atActivity, headerWidth, activityWidth } = this.state

    if (!post && !pending) {
      return <NotFound />
    }

    if (pending) {
      return <Loading />
    }

    const scrollToBottom = () => {
      const detail = document.getElementById(DETAIL_COLUMN_ID)
      detail.scrollTop = detail.scrollHeight
    }

    const headerStyle = {
      width: headerWidth + 'px'
    }
    const activityStyle = {
      width: activityWidth + 'px',
      marginTop: STICKY_HEADER_SCROLL_OFFSET + 'px'
    }

    return <div styleName='post' ref={this.setHeaderStateFromDOM}>
      <ScrollListener elementId={DETAIL_COLUMN_ID}
        onScroll={this.handleScroll} />
      <WrappedPostHeader {...this.props} />
      {atHeader && <div styleName='header-sticky' style={headerStyle}>
        <WrappedPostHeader {...this.props} />
      </div>}
      <PostImage postId={post.id} styleName='image' linked />
      <PostTags tags={post.tags} />
      <PostBody title={post.title}
        id={post.id}
        details={post.details}
        linkPreview={post.linkPreview}
        slug={slug}
        expanded
        styleName='body'
        fileAttachments={post.fileAttachments} />
      <div styleName='activity-header' ref={this.setActivityStateFromDOM}>ACTIVITY</div>
      <PostFooter id={post.id}
        commenters={post.commenters}
        commentersTotal={post.commentersTotal}
        votesTotal={post.votesTotal}
        myVote={post.myVote} />
      {atActivity && <div styleName='activity-sticky' style={activityStyle}>
        <div styleName='activity-header'>ACTIVITY</div>
        <PostFooter id={post.id}
          commenters={post.commenters}
          commentersTotal={post.commentersTotal}
          votesTotal={post.votesTotal}
          myVote={post.myVote} />
      </div>}
      <Comments postId={post.id} slug={slug} scrollToBottom={scrollToBottom} />
      <SocketSubscriber type='post' id={post.id} />
    </div>
  }
}

function WrappedPostHeader ({post, showCommunity, onClose, slug}) {
  return <PostHeader creator={post.creator}
    date={post.createdAt}
    type={post.type}
    communities={post.communities}
    showCommunity={showCommunity}
    close={onClose}
    slug={slug}
    pinned={post.pinned}
    topics={post.topics}
    styleName='header'
    id={post.id}
    topicsOnNewline
    announcement={post.announcement}
  />
}

export function PostTags ({ tags, slug }) {
  if (isEmpty(tags)) return null

  return <div styleName='tags'>
    {tags.map(tag => <Link styleName='tag' to={tagUrl(tag, slug)} key={tag}>
      #{tag}
    </Link>)}
  </div>
}
