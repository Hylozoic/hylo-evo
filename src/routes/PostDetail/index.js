import PostDetail from './PostDetail'
import connector from './PostDetail.connector'
export default connector(PostDetail)
