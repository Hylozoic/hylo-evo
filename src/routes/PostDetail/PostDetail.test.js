import PostDetail, { PostTags } from './PostDetail'
import { shallow } from 'enzyme'
import React from 'react'
import { PostHeader, PostImage, PostBody, PostFooter } from 'components/PostCard/component'
import { getPost } from './PostDetail.connector'
import orm from 'store/models'

describe('PostDetail', () => {
  it('renders correctly', () => {
    const context = 'a project'
    const imageUrl = 'foo.jpg'
    const tags = ['singing', 'dancing']
    const details = 'the body of the post'
    const votesTotal = 7
    const slug = 'foo'

    const post = {
      id: '91',
      context,
      imageUrl,
      tags,
      details,
      votesTotal
    }

    const wrapper = shallow(<PostDetail post={post} slug={slug} />)
    expect(wrapper.find(PostHeader).length).toEqual(1)
    expect(wrapper.find(PostHeader).prop('context')).toEqual(context)
    expect(wrapper.find(PostImage).length).toEqual(1)
    expect(wrapper.find(PostImage).prop('imageUrl')).toEqual(imageUrl)
    expect(wrapper.find(PostTags).length).toEqual(1)
    expect(wrapper.find(PostTags).prop('tags')).toEqual(tags)
    expect(wrapper.find(PostBody).length).toEqual(1)
    expect(wrapper.find(PostBody).prop('details')).toEqual(details)
    expect(wrapper.find(PostFooter).length).toEqual(1)
    expect(wrapper.find(PostFooter).prop('votesTotal')).toEqual(votesTotal)
    expect(wrapper.find('Connect(Comments)').length).toEqual(1)
    expect(wrapper.find('Connect(Comments)').prop('postId')).toEqual(post.id)
    expect(wrapper.find('Connect(Comments)').prop('slug')).toEqual(slug)
  })
})

describe('getPost', () => {
  it("returns null if post doesn't exist", () => {
    const session = orm.session(orm.getEmptyState())
    expect(getPost(session.state, {match: {params: {postId: '1'}}})).toEqual(null)
  })

  it('returns the post with creator, commenters and communities', () => {
    const session = orm.session(orm.getEmptyState())
    const { Post, Person, Community } = session;
    [{
      model: Community, attrs: {id: '1', slug: 'foo'}
    },
    {
      model: Community, attrs: {id: '2', slug: 'bar'}
    },
    {
      model: Community, attrs: {id: '3', slug: 'baz'}
    },
    {
      model: Person, attrs: {id: '1', name: 'Jack'}
    },
    {
      model: Person, attrs: {id: '2'}
    },
    {
      model: Person, attrs: {id: '3', name: 'Sue'}
    },
    {
      model: Post, attrs: {id: '1', title: 'Hay', commenters: ['1'], creator: '3', communities: ['1', '2']}
    },
    {
      model: Post, attrs: {id: '2', commenters: ['2'], creator: '1', communities: ['2', '3']}
    }].forEach(spec => spec.model.create(spec.attrs))

    const result = (getPost({orm: session.state}, {match: {params: {postId: '1'}}}))
    expect(result.title).toEqual('Hay')
    expect(result.creator.name).toEqual('Sue')
    expect(result.commenters.length).toEqual(1)
    expect(result.commenters[0].name).toEqual('Jack')
    expect(result.communities.length).toEqual(2)
    expect(result.communities.map(c => c.slug)).toEqual(['foo', 'bar'])
  })
})
