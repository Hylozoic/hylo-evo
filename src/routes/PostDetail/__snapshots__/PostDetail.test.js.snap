// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetail renders correctly 1`] = `
<div
  data-stylename="post"
>
  <ScrollListener
    elementId="detail-column"
    onScroll={[Function]}
  />
  <Connect(PostHeader)
    data-stylename="header"
    details="the body of the post"
    id="91"
    imageUrl="foo.jpg"
    members={Array []}
    myVote={true}
    routeParams={
      Object {
        "slug": "foo",
      }
    }
    tags={
      Array [
        "singing",
        "dancing",
      ]
    }
    topicsOnNewline={true}
    votesTotal={7}
  />
  <Connect(PostImage)
    data-stylename="image"
    linked={true}
    postId="91"
  />
  <PostTags
    tags={
      Array [
        "singing",
        "dancing",
      ]
    }
  />
  <PostBody
    data-stylename="body"
    expanded={true}
    post={
      Object {
        "details": "the body of the post",
        "id": "91",
        "imageUrl": "foo.jpg",
        "members": Array [],
        "myVote": true,
        "tags": Array [
          "singing",
          "dancing",
        ],
        "votesTotal": 7,
      }
    }
    slug="foo"
  />
  <PostCommunities
    showBottomBorder={true}
    slug="foo"
  />
  <div
    data-stylename="activity-header"
  >
    ACTIVITY
  </div>
  <PostFooter
    details="the body of the post"
    id="91"
    imageUrl="foo.jpg"
    members={Array []}
    myVote={true}
    tags={
      Array [
        "singing",
        "dancing",
      ]
    }
    voteOnPost={[Function]}
    votesTotal={7}
  />
  <Connect(Comments)
    postId="91"
    scrollToBottom={[Function]}
    slug="foo"
  />
  <Connect(SocketSubscriber)
    id="91"
    type="post"
  />
</div>
`;

exports[`ProjectContributions disables button on invalid amount 1`] = `
<ProjectContributions
  postId={123}
  processStripeToken={[Function]}
  totalContributions={321}
>
  <div
    data-stylename="project-contributions"
  >
    <div>
      <div
        data-stylename="amount-row"
      >
        <span
          data-stylename="amount-label"
        >
          Amount
        </span>
        <TextInput
          inputRef={[Function]}
          noClearButton={true}
          onChange={[Function]}
          value="$12fc"
        >
          <div
            data-stylename="wrapper"
          >
            <input
              data-stylename="input"
              onChange={[Function]}
              onKeyDown={[Function]}
              value="$12fc"
            />
          </div>
        </TextInput>
      </div>
      <ReactStripeCheckout
        ComponentClass="span"
        amount={NaN}
        className="StripeCheckout"
        disabled={true}
        label="Pay With Card"
        locale="auto"
        name="Contributing Via Stripe"
        reconfigureOnUpdate={false}
        stripeKey="pk_test_q0UO6c12K9d6oP8DGiLCxr2b"
        token={[Function]}
        triggerEvent="onClick"
      >
        <button
          disabled={true}
          style={
            Object {
              "background": "rgba(0,0,0,0.2)",
              "border": 0,
              "borderRadius": 5,
              "display": "inline-block",
              "overflow": "hidden",
              "padding": 1,
              "textDecoration": "none",
              "userSelect": "none",
            }
          }
        >
          <span
            style={
              Object {
                "background": "#f8f9fa",
                "borderRadius": 4,
                "boxShadow": "inset 0 1px 0 rgba(255,255,255,0.25)",
                "color": "#999",
                "display": "block",
                "fontFamily": "\\"Helvetica Neue\\",Helvetica,Arial,sans-serif",
                "fontSize": 14,
                "height": 30,
                "lineHeight": "30px",
                "padding": "0 12px",
                "position": "relative",
                "textShadow": "0 1px 0 rgba(255,255,255,0.5)",
              }
            }
          >
            Pay With Card
          </span>
        </button>
      </ReactStripeCheckout>
      <Button
        color="gray"
        data-stylename="cancel-button"
        label="Cancel"
        narrow={true}
        onClick={[Function]}
        small={true}
      >
        <div
          data-stylename="button gray narrow small"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          Cancel
        </div>
      </Button>
    </div>
    <div
      data-stylename="project-contributions-total"
    >
      Contributions so far: $
      321
    </div>
  </div>
</ProjectContributions>
`;

exports[`ProjectContributions renders correctly 1`] = `
<div
  data-stylename="project-contributions"
>
  <Button
    color="green"
    label="Contribute"
    narrow={true}
    onClick={[Function]}
    small={true}
  />
  <div
    data-stylename="project-contributions-total"
  >
    Contributions so far: $
    321
  </div>
</div>
`;

exports[`ProjectContributions renders error message 1`] = `
<ProjectContributions
  postId={123}
  processStripeToken={[Function]}
  totalContributions={321}
>
  <div
    data-stylename="project-contributions"
  >
    <div
      data-stylename="error-notification"
    >
      There was a problem processing your payment. Please check your card details and try again.
    </div>
    <Button
      color="green"
      label="Contribute"
      narrow={true}
      onClick={[Function]}
      small={true}
    >
      <div
        data-stylename="button green narrow small"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Contribute
      </div>
    </Button>
    <div
      data-stylename="project-contributions-total"
    >
      Contributions so far: $
      321
    </div>
  </div>
</ProjectContributions>
`;

exports[`ProjectContributions renders received message 1`] = `
<ProjectContributions
  postId={123}
  processStripeToken={[Function]}
  totalContributions={321}
>
  <div
    data-stylename="project-contributions"
  >
    <div
      data-stylename="success-notification"
    >
      Thanks for your contribution!
    </div>
    <div
      data-stylename="project-contributions-total"
    >
      Contributions so far: $
      321
    </div>
  </div>
</ProjectContributions>
`;
